### ISSUE TITLE
bug(lib.NewInstance): repo lock problems blocking moving forward

### ISSUE THREAD

**Issue by ramfox** [qri-io/qri#1375]
Okay now that I've flipped back to upgrading `qri `to work with the latest `qfs` (which is blocking us from finishing the migration work)

We are running into repo lock issues in two specific circumstances. (We have new repo lock issues now because `qfs` now checks to see if the repo is locked before adding it as a filesystem).

#### run `qri connect` in one terminal and another command in another terminal:
To walk through the issue:
- run qri connect - this creates an instance of qri, which puts a lock on the underlying ipfs repo
- in a new terminal run any qri command `qri ls` for example, you will get a repo lock error
Why?
Let's use `qri list` as an example:
The qri `list` cmd, in the `RunE` section of the cobra command, runs `o.Complete()`. This fills out any options. For the `qri list` command, it fills out the `ListOptions` struct, which includes the `DatasetMethods` field. We use the `f.DatasetMethods()` function, which requires an instance in order to work. This creates a new instance, the instance tries to access the ipfs repo, which has a lock, and we error.
We don't look to see if we should actually be connecting via RPC when we are down in lib, after the instance is created. So we have a bit of a catch22 here, we can't check to see if we can communicate over RPC because we need an instance to do that, but we can't create an instance because the repo is locked!

#### cmd tests where we call two commands in a row using `run.ExecCommand` on the same TestRunner
Prime example of a test that calls multiple commands in a row:
https://github.com/qri-io/qri/blob/b0c585c9784c5af3bf332f913b84d496eac77eab/cmd/save_test.go#L1060-L1084
Looks like we don't release the Instance between running the commands.
Do we have a functional way of releasing the Instance, so we can do it when necessary?

**Comment by ramfox** [comment]
My main... challenge with this, is the only ways I can think of to adjust this would be to either roll back the change on qfs that checks for the lock... or very big refactors. Which just means another block on our road to upgrading ipfs/the qri config

**Comment by b5** [comment]
> We don't look to see if we should actually be connecting via RPC when we are down in lib, after the instance is created. 

This isn't right. `lib.NewInstance` _does_ check the configured RPC port (if config.RPC is enabled), and returns early with a sparse `*lib.Instance` if it can connect over RPC:

https://github.com/qri-io/qri/blob/b0c585c9784c5af3bf332f913b84d496eac77eab/lib/lib.go#L328-L338

In that early return the instance is created, but it only has an RPC pointer, and does no repo construction. Something else must be going on. Are we calling things within init that we shouldn't be?

**Comment by b5** [comment]
Spent more time thinking about this. The RPC issue is only part of the problem you articulated @ramfox. We can use the early-return trick to carve around the issue of only grabbing the repo lock once per life of a process. Moving IPFS lock handling into `qfs` shows there are lots of situations where having the lifespan of a repo lock match the lifespan of the process isn't ideal.

The second example of two successive `cmd` tests shows it well. These are modeled as integration tests, and a more realistic modeling of real-world use would create and destroy the `lib.Instance` on each successive call to `runner.MustExec`. Not surprised to hear this is where you're seeing problems first. To fix, we need to address your question of how we release the lock, but I'd rephrase the question: we need better control over the lifespan of resources, including how to release, but also _when_ to release.

The right way to do this in go is with contexts.

**Comment by ramfox** [comment]
PHEW. So glad I misunderstood the first issue! turns out it was a local issue w/ my config and not setting the rpc port correct :)

**Comment by ramfox** [comment]
Okay, so right now when we create a `lib.NewInstance` in the cmd package, we use the context that was passed in/created by the `NewQriCommand` function (called during `root.Execute()` in production, and called when creating a `TestRunner` for the cmd tests. 

This `TestRunner` is our problem child. We create one test runner for multiple command executions, so the context is bound to the TestRunner, not the command executing.

When we run a command, we
1) Complete the command Options - creating a `lib.NewInstance` in the process when we try to get the lib methods
2) Validate the options (when necessary)
3) Run the command

The `lib.NewInstance` needs a context that can be canceled. If we bind the context to the scope of the `RunE` function, then the instance will only exist for the duration of that command execution.

This would also mean adjusting each command to create a context with cancel and pass it down to the `Complete` function. 

The `cmd.Init()` function would also have to be adjusted to taking a context.

The `QriOptions` technically has a `ctx` field, but this has been marked for deprecation in favor of passing a context down.

Tried this for one of our tests `TestReadmeFromFile`, we fall into the same error as we did when testing in fsi: you still need a bit of time for the resources to be released (aka `time.Sleep`) or there are still repo lock conflicts. 

Should we be propagating up this `Done()` pattern, so we can report when all the resources have been released & hold off of the completion of the command until that point?

In `run.MustExec`, if you follow the stack down, we eventually call `cobra.ExecuteC`. We can switch over to calling `cobra.ExecuteContext`, which wraps the execute in a context, and then wait on the Done channel of that context before moving forward with the next `run.MustExec`?

**Comment by b5** [comment]
> The lib.NewInstance needs a context that can be canceled.

Yesyes 💯.

> Should we be propagating up this Done() pattern, so we can report when all the resources have been released & hold off of the completion of the command until that point?

Yes absolutely. I think lib.Instance itself should have a `Done()` method that replicates the same fan-in pattern we're experimenting with on `muxfs`: systems that compose numerous `Done` implementations should compose & publish on a done channel when all others are done.

> In run.MustExec, if you follow the stack down, we eventually call cobra.ExecuteC. We can switch over to calling cobra.ExecuteContext, which wraps the execute in a context, and then wait on the Done channel of that context before moving forward with the next run.MustExec?

EXACTLY. each one of those calls to execute a function should be happening in it's own created-then-cancelled context, with the function itself not returning until receiving on `<-cmd.Instance.Done()`.  Something like this:

```go
package cmd

// test_runner_test.go around line 184...

// ExecCommand executes the given command string
func (run *TestRunner) ExecCommand(cmdText string) error {
	// create a context with a timeout so things that don't clean up properly still fail
	ctx, cancel := context.WithTimeout(context.Background(), time.Second * 5)
	
	var doneCh chan struct{}
	// createCommandRunner needs to return a done channel, more on this in a second
	run.CmdR, doneCh = run.CreateCommandRunner(ctx)

	// execute the command
	if err := executeCommand(run.CmdR, cmdText); err != nil {
		cancel()
		return err
	}

	// command succeeded, close the context
	cancel()
	// wait for cleanup
	<- doneCh
	// now we can move on
	return nil
}
```

The big challenge here is plumbing the done channel back up into this function. `NewQriCommand` returns a `*cobra.Command`, which we can't easily modify. I think the solution here is to change what's passed to `NewQriCommand`

**Comment by b5** [comment]
closed by #1387
